### ISSUE TITLE
[Sub-RFC] BackendFeature: Framework-wide Capability Extension

### ISSUE THREAD

**Issue by 1479518308** [pytorch/pytorch#191317]
# [Sub-RFC] BackendFeature: Framework-wide Capability Extension

## TL;DR

Promote the existing `BackendFeature` enum from an inductor-codegen-only abstraction to a framework-wide capability interface. Add a set of framework-level members (starting with `GPU`), introduce a `DeviceInterface.backend_features(device)` method returning the set of features a backend supports, and replace the scattered hardcoded device-name lists across the framework (`GPU_TYPES`, online-softmax, etc.) with `BackendFeature` membership checks. Boolean capabilities ship first; parameterized negotiation is left to a future RFC.

## Motivation

PyTorch has many "does this backend support X" checks that, lacking a first-class capability interface, are written as `device.type in ["cuda", ...]`:

- `torch/_inductor/utils.py` `GPU_TYPES = ["cuda", "mps", "xpu", "mtia"]`
- `torch/_inductor/codegen/common.py` online-softmax check

This binds "capability" to "backend identity": an OOT / PrivateUse1 backend that actually implements a capability cannot advertise it through the existing interface and has to monkeypatch upstream symbols. The same capability ends up duplicated across multiple hand-maintained device-name lists in the framework, drifting and breaking as backends evolve.

`BackendFeature` already exists as Inductor's codegen-backend capability interface (`has_backend_feature` / `V.graph.has_feature` / `get_backend_features(device)`), but it only fires on the inductor codegen path — the eager path never instantiates the `Scheduling` class, and downstream libraries cannot see it. This RFC extends it to a framework-wide interface, reusing the same enum and the same query channel, without introducing a parallel registration mechanism.

## Approach

1. **Extend the `BackendFeature` enum**: keep the existing inductor-codegen members (FOREACH/BUCKETIZE/SCAN/SORT/TRITON_TEMPLATES/...), add framework-level members (starting with `GPU`; candidates for follow-up: `ONLINE_SOFTMAX` / `FLASH_LINEAR_ATTENTION` / `BF16`).
2. **`DeviceInterface.backend_features(device)`**: new method returning the set of `BackendFeature` members this backend supports (`set[BackendFeature]` or `frozenset`). Base class defaults to an empty set; unknown backends stay conservative. In-tree backends override to fill in the members they actually support.
3. **Derive `GPU_TYPES`**: from the registry, filtering `BackendFeature.GPU in iface.backend_features(None)`.
4. **Migrate the scattered device-name lists** one by one to `BackendFeature` membership checks.
5. **Boolean first, rich negotiation later**: v1 ships yes/no capability advertisement; parameterized/conditional capabilities (e.g. SDPA's head_dim/mask_type constraints) are left to v2.

## Components

- **`BackendFeature` (extended)** — existing enum in `torch/_inductor/codegen/common.py`, with new framework-level members.
- **`DeviceInterface.backend_features(device)`** — new method on `DeviceInterface` in `torch/_dynamo/device_interface.py`; base default empty, in-tree interfaces override.
- **`GPU_TYPES` derivation** — `torch/_inductor/utils.py` becomes a late-binding shim derived from `get_registered_device_interfaces()`.
- **Migration points** — `torch/_inductor/codegen/common.py` (online-softmax), etc.
- **`has_backend_feature` / `get_backend_features`** — the existing inductor consumption channel, reused as-is; no parallel query function is introduced.

### 1. `BackendFeature` enum extension

`torch/_inductor/codegen/common.py`:

```python
class BackendFeature(Enum):
    # —— original Inductor codegen capabilities (migrated from common.py) ——
    FOREACH = auto()
    BUCKETIZE = auto()
    INPLACE_BUFFERS = auto()
    MASKED_SCATTER_WITH_INDEX = auto()
    SCAN = auto()
    SORT = auto()
    TUPLE_REDUCTION = auto()
    PREFER_STORE_LOOP_ORDER = auto()
    TRITON_TEMPLATES = auto()
    REDUCE_TO_SINGLE_ELEMENT = auto()
    # —— framework-level (incl. eager) capabilities ——
    GPU = auto()
    # candidates for follow-up:
    # ONLINE_SOFTMAX = auto()
    # FLASH_LINEAR_ATTENTION = auto()
    # BF16 = auto()
```

The original inductor-codegen members are unchanged; only `GPU` and other framework-level members are appended. The two-tier comments clearly separate inductor-codegen-only from framework-level, for easier governance going forward.

### 2. `DeviceInterface.backend_features(device)`

`torch/_dynamo/device_interface.py`:

```python
class DeviceInterface:
    # ... existing methods ...

    @staticmethod
    def backend_features(device: torch.types.Device = None) -> set[BackendFeature]:
        """Returns the set of BackendFeature members this backend supports.
        Base default is empty; unknown backends stay conservative. In-tree
        backends override to fill in the members they actually support."""
        return set()
```

In-tree backend override examples:

```python
class CudaInterface(DeviceInterface):
    @staticmethod
    def backend_features(device=None) -> set[BackendFeature]:
        return {
            BackendFeature.FOREACH,
            BackendFeature.BUCKETIZE,
            BackendFeature.SCAN,
            BackendFeature.SORT,
            BackendFeature.TRITON_TEMPLATES,
            # ...
            BackendFeature.GPU,
        }


class CpuInterface(DeviceInterface):
    @staticmethod
    def backend_features(device=None) -> set[BackendFeature]:
        return {BackendFeature.FOREACH, BackendFeature.SORT, ...}  # no GPU


class MtiaInterface(DeviceInterface):
    @staticmethod
    def backend_features(device=None) -> set[BackendFeature]:
        return {BackendFeature.GPU, ...}  # has GPU
```

### 3. `GPU_TYPES` derivation

`torch/_inductor/utils.py`:

```python
GPU_TYPES = [
    name
    for name, iface in get_registered_device_interfaces()
    if ":" not in name                       # only device-type names, skip "cuda:0"
    and BackendFeature.GPU in iface.backend_features(None)
]
```

`GPU_TYPES` becomes a module-level `__getattr__` late-binding shim (consistent with RFC 1), backed by a private `_gpu_types()` that iterates `get_registered_device_interfaces()`; no new public function is added. Backends that do not register `BackendFeature.GPU` automatically stay out of `GPU_TYPES`, equivalent to "not a GPU-class device".

### 4. Migrating the scattered device-name lists

| Current hardcode | After migration |
|---|---|
| `torch/_inductor/utils.py` `GPU_TYPES = ["cuda", "mps", "xpu", "mtia"]` | derived above |
| `torch/_inductor/codegen/common.py` `prepare_softmax_extra_check` `device_type in ["cuda", "xpu"] and ... == "triton"` | new `ONLINE_SOFTMAX` member; inductor may still keep a per-inductor-backend secondary switch (see §5) |

### 5. Inductor-codegen-only vs framework-level

The original inductor-codegen members of `BackendFeature` (FOREACH/BUCKETIZE/SCAN/SORT/TRITON_TEMPLATES/...) continue to serve as the inductor codegen-backend capability interface, consumed by `has_backend_feature` / `V.graph.has_feature` / `get_backend_features(device)`. The framework-level members (`GPU` / ...) are consumed by eager / dispatch / downstream libraries via `iface.backend_features(device)`.

- Inductor-codegen-only capabilities (e.g. `TRITON_TEMPLATES`) stay on the inductor consumption channel.
- Cross-stack capabilities (e.g. `GPU`) are read directly from `backend_features(device)` by eager / downstream libraries; inductor may also read the same set, removing the need to maintain a second list.
- Both share the same enum; they are not mutually exclusive and coexist during the migration.

## Tasks

```
### RFC: BackendFeature Framework-wide Capability Extension
- [ ] Extend `BackendFeature` enum: add `GPU` member (framework-level tier)
- [ ] `DeviceInterface.backend_features(device)` base method (default empty) + in-tree interface overrides
- [ ] `GPU_TYPES` becomes a late-binding shim derived from `get_registered_device_interfaces()`
- [ ] Migrate inductor online-softmax check (add `ONLINE_SOFTMAX` member, coexist with inductor secondary switch)
- [ ] Unit tests: each in-tree backend's `backend_features` return value matches real capability; `GPU_TYPES` derivation zero-regression; migration points zero-regression
- [ ] Docs: `BackendFeature` enum docstring marking the framework-level tier; `DeviceInterface.backend_features` public docstring; downstream library migration guide
```

## Open Questions

1. **Capability vocabulary governance**: does adding a framework-level member require an RFC, or is PR + review enough? Leaning: new members go through PR review; semantic changes (rename / meaning change of a member) go through an RFC.

2. **`backend_features` return type**: `set[BackendFeature]` or `frozenset[BackendFeature]`? Leaning: `frozenset`, emphasizing immutability and cacheability.

3. **Timing of parameterized negotiation `supports(op, ...)`**: ultimately `supports(op=aten.scaled_dot_product_attention, device, dtype, head_dim, is_causal, mask_type)` is needed. v1 ships booleans only. When does `supports` land? Does it need a canonical op-capability vocabulary first? Leaning: not implemented in v1; real `supports` implementation deferred to a separate RFC (op-level capability registry, dispatcher negotiation).

4. **Gated-delta attention double gap**: this scenario lacks both (a) a capability interface and (b) a backend-neutral canonical op (e.g. `aten.gated_delta_attention`). This RFC only addresses (a); (b) requires a separate canonical-op proposal. This RFC marks it as a dependency but out of scope.

5. **Physical location of `BackendFeature`**: the enum currently lives in `torch/_inductor/codegen/common.py`. Once promoted to framework-level, should it move to `torch/_dynamo/device_interface.py` or a standalone module, to avoid the eager path importing inductor? Leaning: move to `torch._dynamo.device_interface` (co-located with `DeviceInterface`); inductor re-exports from there, so eager does not pull in inductor just to query capabilities.

## Relationship to the other sub-RFCs

- **RFC 1 (#189135)**: adds `is_gpu()` / `is_triton_capable()` etc. boolean capability bits on `DeviceInterface`. This RFC folds that kind of bit into the `BackendFeature` set returned by `backend_features()`; `is_gpu()` can become a convenience wrapper over `BackendFeature.GPU in self.backend_features(device)`. The two are equivalent and serve as fallbacks for each other. The semantic root of `GPU_TYPES` moves from a hardcoded list to the `BackendFeature.GPU` membership bit.

- **RFC 2 (#189136) / RFC 3 (#189137)**: Dynamo/Inductor consume capabilities. Once this RFC provides a unified capability source, dynamo/inductor's `is_gpu()` and similar queries can be changed to read `backend_features(device)` as a single source; this RFC does not force that migration, preserving RFC 1's fast path.

- **RFC 4 (#191314, Convention-based Lazy Registration)**: the convention path lets downstream register `DeviceInterface` without importing dynamo. `backend_features()` is a method on `DeviceInterface`; downstream that registers an interface via the convention path overrides `backend_features()` at the same time to advertise capabilities. The two RFCs compose orthogonally.

### In scope

- `BackendFeature` enum extension (new framework-level members, starting with `GPU`);
- `DeviceInterface.backend_features(device)` method + in-tree overrides;
- `GPU_TYPES` derivation + migration of scattered device-name lists;
- Unit tests + docs.

### Out of scope

- Real implementation of parameterized negotiation `supports(op, ...)` (v2, separate RFC);
- New canonical ops (e.g. `aten.gated_delta_attention`) (the (b) side of the double gap, separate proposal);
- Deprecating RFC 1's `DeviceInterface.is_gpu()` and other boolean methods (kept as convenience wrappers over `BackendFeature.GPU`);
- Actual migration of downstream libraries (Transformers/Qwen) (provide a guide, but do not land it for them).

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo @azahed98 @NmomoN @mengpenghui @fwenguang @cdzhan @1274085042 @PHLens @albanD
